source/acquaman/BioXAS/BioXASXASScanConfiguration.cpp

```diff
@@ -0,0 +1,212 @@
+#include "BioXASXASScanConfiguration.h"
+
+#include "acquaman/BioXAS/BioXASXASScanActionController.h"
+#include "beamline/BioXAS/BioXASBeamline.h"
+#include "dataman/AMScanAxisEXAFSRegion.h"
+#include "ui/BioXAS/BioXASXASScanConfigurationEditor.h"
+#include "util/AMEnergyToKSpaceCalculator.h"
+#include "util/AMPeriodicTable.h"
+#include "util/AMElement.h"
+#include "util/AMAbsorptionEdge.h"
+
+#include <QStringBuilder>
+
+BioXASXASScanConfiguration::BioXASXASScanConfiguration(QObject *parent) :
+	AMGenericStepScanConfiguration(parent), BioXASScanConfiguration()
+{
+	setName(description());
+	setUserScanName(description());
+}
+
+BioXASXASScanConfiguration::BioXASXASScanConfiguration(const BioXASXASScanConfiguration &original) :
+	AMGenericStepScanConfiguration(original), BioXASScanConfiguration()
+{
+	setEdge(original.edge());
+	setEnergy(original.energy());
+}
+
+BioXASXASScanConfiguration::~BioXASXASScanConfiguration()
+{
+
+}
+
+AMScanConfiguration* BioXASXASScanConfiguration::createCopy() const
+{
+	AMScanConfiguration *configuration = new BioXASXASScanConfiguration(*this);
+	configuration->dissociateFromDb(true);
+	return configuration;
+}
+
+AMScanController* BioXASXASScanConfiguration::createController()
+{
+	AMScanActionController *controller = new BioXASXASScanActionController(this);
+	controller->buildScanController();
+
+	return controller;
+}
+
+AMScanConfigurationView* BioXASXASScanConfiguration::createView()
+{
+	return new BioXASXASScanConfigurationEditor(this);
+}
+
+void BioXASXASScanConfiguration::setControl(AMControlInfo newInfo)
+{
+	AMGenericStepScanConfiguration::setControl(0, newInfo);
+	setupDefaultXANESRegions();
+}
+
+void BioXASXASScanConfiguration::removeControl()
+{
+	AMGenericStepScanConfiguration::removeControl(0);
+}
+
+void BioXASXASScanConfiguration::clearRegions()
+{
+	for (int axisIndex = 0, axisCount = scanAxes_.count(); axisIndex < axisCount; axisIndex++) {
+		AMScanAxis *scanAxis = scanAxisAt(axisIndex);
+
+		if (scanAxis)
+			foreach (AMScanAxisRegion *region, scanAxis->regions().toList())
+				removeRegion(axisIndex, region);
+	}
+}
+
+void BioXASXASScanConfiguration::setupDefaultXANESRegions()
+{
+	// Clear previous regions.
+
+	clearRegions();
+
+	// Create new region and add to the list of regions for the first scan axis.
+
+	if (scanAxisAt(0)) {
+		AMScanAxisEXAFSRegion *region = createDefaultXANESRegion(energy());
+		addRegion(0, 0, region);
+	}
+}
+
+void BioXASXASScanConfiguration::setupDefaultEXAFSRegions()
+{
+	// Clear previous regions.
+
+	clearRegions();
+
+	// Create new regions and add to the list of regions for the first scan axis.
+
+	if (scanAxisAt(0)) {
```

Comment:
It's tough to say. I kind of sidestep this issue on BioXAS by initializing all XAS scan configurations in the app controller--there's never an XAS configuration displayed that doesn't have an appropriate control and regions already provided. It might be useful to always have regions available, even if they're default "nonsense" regions. However, I think it makes more conceptual sense to say that regions are tied more closely to the scanned control than they are inherent properties of scan configurations. But it's kind of a toss-up either way.